crates/ironrdp-rdpdr/src/pdu/esc.rs

```diff
@@ -1309,3 +1311,138 @@ impl rpce::HeaderlessEncode for TransmitReturn {
+    fn decode(src: &mut ReadCursor<'_>) -> PduResult<Self> {
+        let mut index = 0;
+        let mut handle = ScardHandle::decode_ptr(src, &mut index)?;
+        ensure_size!(in: src, size: size_of::<u32>() * 3);
+        let reader_names_is_null = src.read_u32() == 1;
+        let reader_length = src.read_u32();
+        let atr_length = src.read_u32();
+        handle.decode_value(src)?;
+        Ok(Self {
+            handle,
+            reader_names_is_null,
+            reader_length,
+            atr_length,
+        })
+    }
+}
+
+/// [2.2.3.10 Status_Return]
+///
+/// [2.2.3.10 Status_Return]: https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpesc/987c1358-ad6b-4c8e-88e1-06210c28a66f
+#[derive(Debug)]
+pub struct StatusReturn {
+    pub return_code: ReturnCode,
+    pub reader_names: Vec<String>,
+    pub state: CardState,
+    pub protocol: CardProtocol,
+    pub atr: [u8; 32],
+    pub atr_length: u32,
+
+    pub encoding: CharacterSet,
+}
+
+impl StatusReturn {
+    const NAME: &'static str = "Status_Return";
+
+    pub fn new(
+        return_code: ReturnCode,
+        reader_names: Vec<String>,
+        state: CardState,
+        protocol: CardProtocol,
+        atr: [u8; 32],
+        atr_length: u32,
+        encoding: CharacterSet,
+    ) -> rpce::Pdu<Self> {
+        rpce::Pdu(Self {
+            return_code,
+            reader_names,
+            state,
+            protocol,
+            atr,
+            atr_length,
+            encoding,
+        })
+    }
+}
+
+impl rpce::HeaderlessEncode for StatusReturn {
+    fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
+        ensure_size!(in: dst, size: self.size());
+        dst.write_u32(self.return_code.into());
+        let mut index = 0;
+        let reader_names_length: u32 = cast_length!(
+            "StatusReturn",
+            "reader_names_length",
+            encoded_multistring_len(&self.reader_names, self.encoding)
+        )?;
+        ndr::encode_ptr(Some(reader_names_length), &mut index, dst)?;
+        dst.write_u32(self.state.into());
+        dst.write_u32(self.protocol.bits());
+        dst.write_slice(&self.atr);
+        dst.write_u32(self.atr_length);
+        dst.write_u32(reader_names_length);
+        write_multistring_to_cursor(dst, &self.reader_names, self.encoding)?;
+        Ok(())
+    }
+
+    fn name(&self) -> &'static str {
+        Self::NAME
+    }
+
+    fn size(&self) -> usize {
+        size_of::<u32>() * 5 // dst.write_u32(self.return_code.into()); dst.write_u32(self.state.into()); dst.write_u32(self.protocol.bits()); dst.write_slice(&self.atr); dst.write_u32(self.atr_length);
+        + ndr::ptr_size(true) // ndr::encode_ptr(Some(reader_names_length), &mut index, dst)?;
+        + self.atr.len() // dst.write_slice(&self.atr);
+        + encoded_multistring_len(&self.reader_names, self.encoding) // write_multistring_to_cursor(dst, &self.reader_names, self.encoding)?;
+    }
```

Comment:
@CBenoit I know we've gone in circles on this, I think in these particular cases for PDUs which require this kind of tricky ndr encoding, this kind of notation makes the most sense. (I use this to help me write the function in the first place).

MS-RDPESC is something of an exception in this respect, the sizing isn't as straightforward to follow from the docs (e.g. https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpesc/987c1358-ad6b-4c8e-88e1-06210c28a66f) as compared to the rest of MS-RDPDR (e.g. https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/a00fbce4-95bb-4e15-8182-be2b5ef9076a).

IMO I should use this style in these instances, and then switch to what we discussed with @pacmancoder once I get back to MS-RDPDR proper.

Reply:
Fair point, you’re right, let’s use this style for MS-RDPESC